scater-calculate-qc-metrics.R

```diff
@@ -0,0 +1,107 @@
+suppressPackageStartupMessages(require(optparse))
+
+# Load common functions
+suppressPackageStartupMessages(require(workflowscriptscommon))
+
+# parse options
+option_list = list(
+  make_option(
+    c("-i", "--input-object-file"),
+    action = "store",
+    default = NA,
+    type = 'character',
+    help = "File name in which a serialized R SingleCellExperiment object where object matrix found"
+  ),
+  make_option(
+    c("-e", "--exprs-values"),
+    action = "store",
+    default = 'counts',
+    type = 'character',
+    help= "character string indicating which slot of the assayData from the 'SingleCellExperiment' object should be used to compute log-transformed expression values. Valid options are 'counts', 'tpm', 'cpm' and 'fpkm'. Defaults to the first available value of the options in the order shown."
+  ),
+  make_option(
+    c("-f", "--feature-controls"),
+    action = "store",
+    default = NULL,
+    type = 'character',
+    help = "comma-separated list of files with one feature per line. A named list is created containing one or more vectors to identify feature controls (for example, ERCC spike-in genes, mitochondrial genes, etc)"
+  ),
+  make_option(
+    c("-c", "--cell-controls"),
+    action = "store",
+    default = NULL,
+    type = 'character',
+    help = "a character vector of cell (sample) names, or a logical vector, or a numeric vector of indices used to identify cell controls (for example, blank wells or bulk controls)."
+  ),
+  make_option(
+    c("-n", "--nmads"),
+    action = "store",
+    default = 5,
+    type = 'numeric',
+    help = "numeric scalar giving the number of median absolute deviations to be used to flag potentially problematic cells based on total counts (total number of counts for the cell, or library size) and total_features (number of features with non-zero expression). For total_features, cells are flagged for filtering only if total_features is 'nmads' below the median"
+  ), 
+  make_option(
+    c("-p", "--pct-feature-controls-threshold"),
+    action = "store",
+    default = 80,
+    type = 'numeric',
+    help = "numeric scalar giving a threshold for percentage of expression values accounted for by feature controls. Used as to flag cells that may be filtered based on high percentage of expression from feature controls."
+  ),
+  make_option(
+    c("-o", "--output-object-file"),
+    action = "store",
+    default = NA,
+    type = 'character',
+    help = "File name in which to store serialized R object of type 'SingleCellExperiment'."
+  )
+)
+
+opt <- wsc_parse_args(option_list, mandatory = c('input_object_file', 'output_object_file'))
+
+# Check parameter values defined
+if ( ! file.exists(opt$input_object_file)){
+  stop((paste('File object or matrix', opt$input_object_file, 'does not exist')))
+}
+
+# Check feature_controls
+if (! is.null(opt$feature_controls) && opt$feature_controls != 'NULL'){
+  if (! file.exists(opt$feature_controls)){
```

Comment:
Remember this could be a commma-separated list of filenames we have to check for existence. There's a helpful function in the common library we can use. Lapply is also a better way to go for the following code. And sorry, I think expecting people to supply controls as numeric values is likely to create difficulties- better to assume these are strings of e.g. spike names.

Suggest the following:

feature_controls_files <- wsc_split_string(opt$feature_controls)

# Check the files one-by-one

for (fcf in feature_control_files){
    if (! file.exists(fcf)){
        stop((paste('Supplied feature_controls file', fcf, 'does not exist')))
    }
}

# Read the list of files into a list of vectors
feature_controls <- lapply(feature_control_files, readLines)

# Give the list names derived from the individual file names
names(feature_controls) <- basename(feature_control_files)